pg_search/src/query/mod.rs

```diff
@@ -558,6 +559,163 @@ impl ExplainFormat for SearchQueryInput {
     }
 }
 
+impl SearchQueryInput {
+    pub unsafe fn from_datum_resilient(
+        datum: pg_sys::Datum,
+        is_null: bool,
+    ) -> Option<SearchQueryInput> {
+        if is_null {
+            return None;
+        }
+
+        if let Some(array_query) = Self::coerce_scalar_array(datum, is_null) {
+            return Some(array_query);
+        }
+
+        let bytes = detoast_datum_bytes(datum);
+
+        match serde_cbor::from_slice::<SearchQueryInput>(&bytes) {
+            Ok(parsed) => Some(parsed),
+            Err(cbor_err) => match serde_json::from_slice::<SearchQueryInput>(&bytes) {
+                Ok(parsed) => Some(parsed),
+                Err(json_err) => panic!(
+                    "failed to decode SearchQueryInput (CBOR error: {cbor_err}; JSON error: {json_err}; preview={})",
+                    json_preview(&bytes)
+                ),
+            },
+        }
+    }
+
+    unsafe fn coerce_scalar_array(datum: pg_sys::Datum, is_null: bool) -> Option<SearchQueryInput> {
+        if is_null {
+            return None;
+        }
+
+        let raw_array = pg_sys::pg_detoast_datum_copy(datum.cast_mut_ptr::<pg_sys::varlena>());
```

Comment:
This `detoast_datum_copy` is unnecessary because what `array_from_varlena` really happens to be is `vec_from_arraytype`. You shouldn't be forcing a copy if you may need to `pfree` it later, but also that is handled transparently by using pgrx's implementation of this code via `FromDatum` which will detoast the array to whatever degree necessary and then directly traverse it to create the `Vec`. It's not technically impossible that I'm missing something here, but you use the CBOR representations that are a direct consequence of `derive(PostgresType)` so this should be a perfectly eligible candidate.

I am not entirely sure why you have reimplemented all this here.